src/code/TwoHeadArena.c

```diff
@@ -1,140 +1,135 @@
-    return THA_GetTail((TwoHeadArena*)thga);
-}
-
-Gfx* THGA_AllocStartArray8(TwoHeadGfxArena* thga, u32 count) {
-    return THA_AllocStart((TwoHeadArena*)thga, count * 8);
-}
-
-Gfx* THGA_AllocStart8(TwoHeadGfxArena* thga) {
-    return THGA_AllocStartArray8(thga, 1);
-}
-
-Gfx* THGA_AllocStart8Wrapper(TwoHeadGfxArena* thga) {
-    return THGA_AllocStart8(thga);
-}
-
-Gfx* THGA_AllocEnd(TwoHeadGfxArena* thga, u32 size) {
-    return THA_AllocEnd((TwoHeadArena*)thga, size);
-}
-
-Gfx* THGA_AllocEndArray64(TwoHeadGfxArena* thga, u32 count) {
-    return THGA_AllocEnd(thga, count * 0x40);
-}
-
-Gfx* THGA_AllocEnd64(TwoHeadGfxArena* thga) {
-    return THGA_AllocEnd(thga, 0x40);
-}
-
-Gfx* THGA_AllocEndArray16(TwoHeadGfxArena* thga, u32 count) {
-    return THGA_AllocEnd(thga, count * 0x10);
-}
-
-Gfx* THGA_AllocEnd16(TwoHeadGfxArena* thga) {
-    return THGA_AllocEnd(thga, 0x10);
-}
-
 void* THA_GetHead(TwoHeadArena* tha) {
     return tha->head;
 }
 
-void THA_SetHead(TwoHeadArena* tha, void* start) {
-    tha->head = start;
+void THA_SetHead(TwoHeadArena* tha, void* newHead) {
+    tha->head = newHead;
 }
 
 void* THA_GetTail(TwoHeadArena* tha) {
     return tha->tail;
 }
 
-void* THA_AllocStart(TwoHeadArena* tha, u32 size) {
+/**
+ * Allocates to the head of the Two Head Arena. The allocation will not have any alignment guarantees.
+ */
+void* THA_AllocHead(TwoHeadArena* tha, size_t size) {
     void* start = tha->head;
 
-    tha->head = (void*)((u32)tha->head + size);
+    tha->head = (u8*)tha->head + size;
     return start;
 }
 
-void* THA_AllocStart1(TwoHeadArena* tha) {
-    return THA_AllocStart(tha, 1);
+void* THA_AllocHead1(TwoHeadArena* tha) {
+    return THA_AllocHead(tha, 1);
 }
 
-void* THA_AllocEnd(TwoHeadArena* tha, u32 size) {
-    u32 mask;
+/**
+ * Allocates to the tail end of the Two Head Arena. The allocation will be aligned based on the size of the allocation.
+ * For all allocations greater than 16-bytes, they will be aligned to 16-bytes. Otherwise, the alignment will be the
+ * largest power of 2 for which the size is a multiple, in order to accommodate the alignment requirements of any data
+ * types that can fit within the allocation.
+ */
+void* THA_AllocTail(TwoHeadArena* tha, size_t size) {
+    uintptr_t mask;
 
     if (size == 8) {
-        mask = ~7;
+        // Align 8 for multiples of 8
+        mask = ALIGN_MASK(8);
     } else if (size == 4 || size == 12) {
-        mask = ~3;
+        // Align 4 for multiples of 4
+        mask = ALIGN_MASK(4);
     } else if (size == 2 || size == 6 || size == 10 || size == 12 || size == 14) {
-        mask = ~1;
+        // Align 2 for multiples of 2
+        mask = ALIGN_MASK(2);
+    } else if (size >= 0x10) {
+        // Align 0x10 for allocations greater than 0x10
+        mask = ALIGN_MASK(0x10);
     } else {
-        mask = (size >= 0x10) ? ~0xF : 0;
+        //! @bug if size is 1 the computation below will give NULL. The mask for this case would be
```

Comment:
```suggestion
        //! @bug if size is less than 16 and odd or zero the computation below will give NULL. The mask for this case would be
```
I think. Not sure on the wording but my point is it also happens in those cases?

Reply:
seems that way, though I question whether a size less than 1 should be considered a valid argument for the function.